eng/common/scripts/Update-DocsMsToc.ps1

```diff
@@ -0,0 +1,313 @@
+    };
+  }
+
+  $uncategorizedPackages = $packagesForToc.Values.Where({ $_.ServiceName -eq $service -and !(@('client', 'mgmt') -contains $_.Type) })
+  if ($uncategorizedPackages) {
+    foreach ($package in $uncategorizedPackages) {
+      LogWarning "Uncategorized package for service: $service - $($package.Package). Package not onboarded."
+    }
+  }
+
+  $serviceReadmeBaseName = $service.ToLower().Replace(' ', '-')
+  $serviceTocEntry = [PSCustomObject]@{
+    name            = $service;
+    href            = "~/docs-ref-services/{moniker}/$serviceReadmeBaseName.md"
+    landingPageType = 'Service'
+    items           = @($packageItems)
+  }
+
+  $toc += $serviceTocEntry
+}
+
+# Core packages belong under the "Other" node in the ToC
+$otherPackageItems = New-Object -TypeName System.Collections.Generic.List[PSCustomObject]
+$otherPackages = $packagesForToc.Values.Where({ $_.ServiceName -eq 'Other' })
+$otherPackages = $otherPackages | Sort-Object -Property DisplayName
+
+if ($otherPackages) {
+  foreach ($otherPackage in $otherPackages) {
+    $segments = $otherPackage.DisplayName.Split('-').ForEach({ $_.Trim() })
+
+
+    if ($segments.Count -gt 1) {
+      $currentNode = $otherPackageItems
+
+      # Iterate up to the penultimate item in the array so that the final item
+      # in the array can be added as a leaf node. Since the array always has at
+      # least two elements this iteration will cover at least the first element.
+      # e.g. @(0, 1)[0..0] => 0
+      foreach ($segment in $segments[0..($segments.Count - 2)]) {
+        $matchingNode = $currentNode.Where({ $_.name -eq $segment })
+
+        # ToC nodes can be "branches" which contain 0 or more branch
+        # or leaf nodes in an "items" field OR they can be leaf nodes which have
+        # a "children" field which can only contain package names or namespaces.
+        # A node cannot contain both "items" and "children". If a node already
+        # has a "children" field then it is a leaf node and cannot take
+        # additional branch nodes.
+        # Children are added using the `GetClientPackageNode` function
+        if ($matchingNode -and $matchingNode.PSObject.Members.Name -contains "children") {
+          LogWarning "Cannot create nested entry for package $($otherPackage.Package) because Segment `"$segment`" in the DisplayName $($otherPackage.DisplayName) is already a leaf node. Excluding package: $($otherPackage.Package)"
+          $currentNode = $null
+          break
+        }
+
+        if ($matchingNode) {
+          $currentNode = $matchingNode[0].items
+        }
+        else {
+          $newNode = [PSCustomObject]@{
+            name            = $segment
+            landingPageType = 'Service'
+            items           = New-Object -TypeName System.Collections.Generic.List[PSCustomObject]
+          }
+          $currentNode.Add($newNode)
+          $currentNode = $newNode.items
+        }
+      }
+
+      if ($null -ne $currentNode) {
```

Comment:
Is it possible to have currentNode be null? If it is wouldn't that be an error which we should probably fail on?

Reply:
We set `$currentNode` to `$null` when the metadata contains a hypothetical scenario like this: 

| Package | DisplayName | ServiceName | 
| --------- | --------------- | -------------- | 
| @azure/core-client | Core - Client | Other | 
| @azuure/core-xml | Core - Client - XML | Other | 

In this case, we would build a node like this for the first: 

```yaml
- name: Core
  landingPageType: Service
  items: 
    - name: Client 
      landingPageType: Service
      children: 
        - @azure/core-client 
```

When we attempt to iterate in on the second node we reach `Client` and see that it is a leaf node (i.e. `children` instead of `items`). In this case we break out of that loop but need a way to say "Don't set an `items` property on this entry in the ToC, it won't work" ... Setting the node to `$null` is the way to do this. 

If we quit the entire process here then we'd end up stopping the ToC update process and blocking on a package because of bad metadata. If we skip a package here that package will still end up in `Uncategorized Packages` and we'll be able to build the rest of the ToC.

Reply:
>If we skip a package here that package will still end up in Uncategorized Packages and we'll be able to build the rest of the ToC.

I guess I'm missing where this happens.